wgpu-hal/src/metal/command.rs

```diff
@@ -21,6 +21,26 @@ impl Default for super::CommandState {
     }
 }
 
+impl Drop for super::CommandState {
+    fn drop(&mut self) {
+        // Metal asserts when a MTLCommandEncoder is deallocated without a call
+        // to endEncoding. This isn't documented at
+        // https://developer.apple.com/documentation/metal/mtlcommandencoder?language=objc
+        // but it manifests as a crash with the message 'Command encoder released without
+        // endEncoding'. To prevent this, we explicitiy call end_encoding on any
+        // encoders we are tracking. Redundant calls appear to be harmless.
+        if let Some(ref encoder) = self.blit {
+            encoder.end_encoding();
+        }
+        if let Some(ref encoder) = self.render {
+            encoder.end_encoding();
+        }
+        if let Some(ref encoder) = self.compute {
+            encoder.end_encoding();
+        }
```

Comment:
This looks an awful lot like the `discard_encoding` method. Could we just call that?

Reply:
It's a bit obfuscated, but yes `discard_encoding` does these steps, and then also sets `raw_cmd_buf` to `None`. That seems fine, and I'll change the `Drop` implementation to call that function instead.